### PR TITLE
Improving testing time 

### PR DESCRIPTION
**Pull request by fehiepsi**
Addresses #362 

### Done
- [x] contrib/test_distributions: 60s -> 0s
- [x] other contrib: 110s -> 55s
- [x] revise test_infer_util + distributions_util, remove tests for standard_gamma in util (this is tested in JAX and in test_distributions): 30s -> 0
- [x] test_distribution: relax mean_var LKJ tests 310s -> 230s
- [x] test_hmc_util 217s -> 67s

### Future PRs
@neerajprad  could you make an attempt at those tests first for which one we need to keep and which one we need to remove? I think we don't need to (e.g. x64 for all tests in test_mcmc) but there are some other strategies to do: e.g. mark skip if in Travis,...
- test_mcmc
- test_mcmc_interface

### CODE REVIEW

**Comment by neerajprad** [comment]
> @neerajprad could you make an attempt at those tests first for which one we need to keep and which one we need to remove? I think we don't need to (e.g. x64 for all tests in test_mcmc) but there are some other strategies to do: e.g. mark skip if in Travis,...

I think we can simply remove many of those tests which are actually just duplicates.

File: test/contrib/test_autoguide.py [hunk 1]
```diff
@@ -15,10 +15,14 @@
 from numpyro.distributions import constraints
 from numpyro.distributions.flows import InverseAutoregressiveTransform
 from numpyro.handlers import substitute
+from numpyro.infer_util import init_to_median
 from numpyro.svi import SVI, elbo
 from numpyro.util import fori_loop
 
 
+init_strategy = partial(init_to_median, num_samples=2)
```
Comment (fehiepsi) [inline test/contrib/test_autoguide.py]: This reduces the time spent to find initial_params a bit. It took a few seconds to call `trace(model).get_trace(*args, **kwargs)` the first time.

File: test/contrib/test_nn.py [hunk 2]
```diff
@@ -97,3 +99,18 @@ def test_masks(input_dim, n_layers, output_dim_multiplier):
                 if mask_skip[idx + jdx * input_dim, kdx]:
                     skip_connections.add(kdx)
             assert_array_equal(list(sorted(skip_connections)), correct)
+
+
+@pytest.mark.parametrize('input_dim', [5, 7])
+def test_masked_dense(input_dim):
```
Comment (fehiepsi) [inline test/contrib/test_nn.py]: I moved the content of `test_masked_dense` to here.

File: test/contrib/test_nn.py [hunk 3]
```diff
@@ -51,9 +53,9 @@ def test_auto_reg_nn(input_dim, hidden_dims, param_dims, skip_connections):
     assert onp.sum(onp.abs(onp.triu(permuted_jac))) == 0.0
 
 
-@pytest.mark.parametrize('input_dim', [5, 8])
-@pytest.mark.parametrize('n_layers', [1, 3])
-@pytest.mark.parametrize('output_dim_multiplier', [1, 4])
+@pytest.mark.parametrize('input_dim', [2, 6])
```
Comment (fehiepsi) [inline test/contrib/test_nn.py]: This test uses many for-loops, which is slow. So I reduced those numbers a bit.

File: test/test_distributions.py [hunk 4]
```diff
@@ -472,7 +475,7 @@ def fn(*args):
 
 @pytest.mark.parametrize('jax_dist, sp_dist, params', CONTINUOUS + DISCRETE)
 def test_mean_var(jax_dist, sp_dist, params):
-    n = 200000
+    n = 20000 if jax_dist is dist.LKJCholesky else 200000
```
Comment (fehiepsi) [inline test/test_distributions.py]: we don't need too many LKJ samples to get good precision.

File: test/test_distributions_util.py [hunk 5]
```diff
@@ -123,60 +122,6 @@ def test_fn(x):
     assert_allclose(jacobian(cumprod)(x), jacobian(test_fn)(x), atol=1e-7)
 
 
-@pytest.mark.parametrize('alpha, shape', [
-    (1., ()),
-    (1., (2,)),
-    (np.array([1., 2.]), ()),
-    (np.array([1., 2.]), (3, 2)),
-])
-def test_standard_gamma_shape(alpha, shape):
```
Comment (fehiepsi) [inline test/test_distributions_util.py]: those tests are covered in JAX and test_distributions.py file.

File: test/test_hmc_util.py [hunk 6]
```diff
@@ -454,10 +458,10 @@ def model(data):
         5,  14,  13,  22,
     ])
 
-    rngs = random.split(random.PRNGKey(1), 12)
+    rngs = random.split(random.PRNGKey(1), 2)
```
Comment (fehiepsi) [inline test/test_hmc_util.py]: it is unnecessary to test the correctness of batched `rngs` with batch=12. I think 2 is enough.

**Comment by fehiepsi** [comment]
@neerajprad Testing time is reduced from 38m to 23m. I am satisfied with the result now so would like to do further tuning when total time goes over 30m again.

**Comment by neerajprad** [comment]
> 19m with this PR.

That's great! Yes, it is best to keep revisiting this periodically.

**Pull request merged**